File: backend/app/schemas/progress.py
from __future__ import annotations

from datetime import datetime

from pydantic import Field

from app.schemas.base import BaseSchema, Score, TimestampedDTO, VersionedDTO
from app.schemas.enums import ProgressStatus, TopicProgressStatus
from app.schemas.ids import ConceptId, CurriculumId, GoalId, ProgressId, TopicId


class NextRecommendedAction(BaseSchema):
    topic_id: TopicId | None = None
    label: str = Field(min_length=1, max_length=220)
    reason: str = Field(min_length=1, max_length=500)


class TopicProgressDTO(BaseSchema):
    topic_id: TopicId
    status: TopicProgressStatus
    completion: Score
    last_score: Score | None = None
    attempt_count: int = Field(default=0, ge=0, le=100)
    completed_at: datetime | None = None
    stuck_count: int = Field(default=0, ge=0, le=20)
    notes: str | None = Field(default=None, max_length=800)


class StuckEventDTO(BaseSchema):
    topic_id: TopicId
    concept_ids: list[ConceptId] = Field(default_factory=list, max_length=12)
    reason: str = Field(min_length=1, max_length=500)
    created_at: datetime


class ProgressStateCreate(BaseSchema):
    goal_id: GoalId
    curriculum_id: CurriculumId
    topic_progress: list[TopicProgressDTO] = Field(min_length=1)
    current_topic_id: TopicId | None = None


class ProgressStateDTO(TimestampedDTO, VersionedDTO):
    progress_state_id: ProgressId
    goal_id: GoalId
    curriculum_id: CurriculumId
    status: ProgressStatus
    overall_completion: Score
    current_topic_id: TopicId | None = None
    topic_progress: list[TopicProgressDTO] = Field(min_length=1)
    weak_concepts: list[ConceptId] = Field(default_factory=list, max_length=20)
    stuck_events: list[StuckEventDTO] = Field(default_factory=list, max_length=50)
    last_activity_at: datetime | None = None
    next_recommended_action: NextRecommendedAction | None = None

File: frontend/lib/types/progress.ts
export type TopicProgressStatus =
  | "not_started"
  | "in_progress"
  | "completed"
  | "stuck"
  | "needs_review";

export type ProgressStatus = "not_started" | "in_progress" | "adaptation_needed" | "completed";

export type NextRecommendedAction = {
  topic_id: string | null;
  label: string;
  reason: string;
};

export type TopicProgressDTO = {
  topic_id: string;
  status: TopicProgressStatus;
  completion: number;
  last_score: number | null;
  attempt_count: number;
  completed_at: string | null;
  stuck_count: number;
  notes: string | null;
};

export type StuckEventDTO = {
  topic_id: string;
  concept_ids: string[];
  reason: string;
  created_at: string;
};

export type ProgressStateDTO = {
  progress_state_id: string;
  goal_id: string;
  curriculum_id: string;
  status: ProgressStatus;
  overall_completion: number;
  current_topic_id: string | null;
  topic_progress: TopicProgressDTO[];
  weak_concepts: string[];
  stuck_events: StuckEventDTO[];
  last_activity_at: string | null;
  next_recommended_action: NextRecommendedAction | null;
  created_at: string;
  updated_at: string;
  schema_version: string;
};
